<a href="https://colab.research.google.com/github/lorenzo-crippa/3M_NLP_ESS_2022/blob/main/Tutorial_One_(R)_Intro_to_String_Manipulation_and_Regular_Expressions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Intro to String Manipulation and Regular Expressions in R

### Douglas Rice

In this notebook, we'll learn about doing standard natural language processing (NLP) tasks in R, and will be introduced to regular expressions. After completing this notebook, you should be familar with:


1.   Manipulating strings in R.
2.   Extracting elements from strings / vectors.
3. Working with Regular Expressions in R





# Working with Strings

Our focus is on "strings", or ordered sequences of characters. Importantly, a string can be short ("Hello!"), long ("supercalafragalisticexpealadotios!"), much longer (like a book), and everything in between. In a word, strings are text. 



## Setup

In [ ]:
install.packages("quanteda")

library(quanteda)
library(tidyverse)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



## Working with Strings

Let's look at a string. These can be specified using double quotes (“) or single quotes (’):

In [ ]:
a_string <- 'Example STRING, with numbers (12, 15 and also 10.2)?!'
a_string

[1] "Example STRING, with numbers (12, 15 and also 10.2)?!"

What happens if you use one of the quotes that *is* inside the string? 

In [ ]:
#my_whoops_string <- "He asked, "Why would you use double quotes""

Ah, a dreaded syntax error! This is because quotes are *special characters*; they signify something in the programming language. 

To get around this conundrum, you can use a \ (backslash) to tell R to *escape* the next character. In the example below, the \" is saying, " is part of the string, not the end of the string.

In [ ]:
my_string_with_double_quotes <- "She answered, \"Convenience, but you never really have to.\""
my_string_with_double_quotes

[1] "She answered, \"Convenience, but you never really have to.\""

If you ever want to see how your string with escape characters displays when printed or (typically) in an editor, use `writeLines()`.

In [ ]:
writeLines(my_string_with_double_quotes)

She answered, "Convenience, but you never really have to."


This can get a little bit confusing. For example, since the backslash character tells R to escape, to indicate an actual backslash character you have to backslash your backslashes:

In [ ]:
a_string_with_backslashes <- "To indicate a backslash, \\, you have to type two: \\\\. Just there, to indicate two backslashes, I had to type four: \\\\\\\\."
a_string_with_backslashes

[1] "To indicate a backslash, \\, you have to type two: \\\\. Just there, to indicate two backslashes, I had to type four: \\\\\\\\."

In [ ]:
writeLines(a_string_with_backslashes)

To indicate a backslash, \, you have to type two: \\. Just there, to indicate two backslashes, I had to type four: \\\\.


## Other Escape Characters

As mentioned above, quotes (single or double) are just one many special escape characters that are used to perform special functions. The most common are two that you’re already used to tapping a keyboard key for without expecting a character to appear on your screen: \t (tab) and \n (newline, or "Enter").

In [ ]:
test_string <- "Hark, a Lark!\t*Tweet Tweet*\n  \nWhere'd it go?"
test_string

[1] "Hark, a Lark!\t*Tweet Tweet*\n  \nWhere'd it go?"

In [ ]:
writeLines(test_string)

Hark, a Lark!	*Tweet Tweet*
  
Where'd it go?


## Sets of Strings 

To this point, we have just looked at a single string. Rarely is that what we are actually interested in manipulating as social scientists, though. More frequently, we have sets of lots and lots of strings. To that end, we can create a vector of strings.

In [ ]:
a_list_of_strings <-  c("Manchester City", "Liverpool", "Chelsea", "Tottenham", "Arsenal")
a_list_of_strings

[1] "Manchester City" "Liverpool"       "Chelsea"         "Tottenham"      
[5] "Arsenal"



Let's load in some other strings to work with. 

In [ ]:
letters_string <- letters
letters_string

[1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q" "r" "s"
[20] "t" "u" "v" "w" "x" "y" "z"

We can create a list from those letters, and call it letters_list

In [ ]:
letters_list <- list(letters_string)
letters_list

[[1]]
 [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q" "r" "s"
[20] "t" "u" "v" "w" "x" "y" "z"

We can also use some built in functionality in the "string" module to create upper-case versions of both the string and the list, and we'll call them `LETTERS`. 

In [ ]:
LETTERS_string <- toupper(letters_string)
LETTERS_string

[1] "A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S"
[20] "T" "U" "V" "W" "X" "Y" "Z"

In [ ]:
LETTERS_list = list(LETTERS_string)
LETTERS_list

[[1]]
 [1] "A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S"
[20] "T" "U" "V" "W" "X" "Y" "Z"

We'll make two different month lists, one with the abbreviation (`month_abb`) and one with the full name (`month_name`)

In [ ]:
month_abb <- c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")
month_abb

[1] "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec"

In [ ]:
month_name <- c("January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December")
month_name

[1] "January"   "February"  "March"     "April"     "May"       "June"     
 [7] "July"      "August"    "September" "October"   "November"  "December"

Let's also bring in two lists of strings for us to work with. The first (`fruit`) is a list of fruits, and the second (`words`) is a list of words. 

In [ ]:
install.packages("ds4psy")
library(ds4psy)
fruits

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependency ‘unikn’


Welcome to ds4psy (v0.8.0)!



[1] "Acai"                            "Ackee"                          
  [3] "Apple"                           "Apricot"                        
  [5] "Avocado"                         "Banana"                         
  [7] "Bell pepper"                     "Bilberry"                       
  [9] "Black sapote"                    "Blackberry"                     
 [11] "Blackcurrant"                    "Blood orange"                   
 [13] "Blueberry"                       "Boysenberry"                    
 [15] "Breadfruit"                      "Buddha's hand (fingered citron)"
 [17] "Cactus pear"                     "Cantaloupe"                     
 [19] "Cherimoya (Custard Apple)"       "Cherry"                         
 [21] "Chico fruit"                     "Chile pepper"                   
 [23] "Clementine"                      "Cloudberry"                     
 [25] "Coconut"                         "Corn kernel"                    
 [27] "Crab apple"                      "Cranberry"                      
 [29] "Cucumber"                        "Currant"                        
 [31] "Damson"                          "Date"                           
 [33] "Dragonfruit"                     "Durian"                         
 [35] "Eggplant"                        "Elderberry"                     
 [37] "Feijoa"                          "Fig"                            
 [39] "Galia melon"                     "Goji berry"                     
 [41] "Gooseberry"                      "Grape"                          
 [43] "Grapefruit"                      "Grewia asiatica phalsa"         
 [45] "Guava"                           "Hala fruit"                     
 [47] "Honeyberry"                      "Honeydew"                       
 [49] "Huckleberry"                     "Jabuticaba"                     
 [51] "Jackfruit"                       "Jalapeno"                       
 [53] "Jambul"                          "Japanese plum"                  
 [55] "Jostaberry"                      "Jujube"                         
 [57] "Juniper berry"                   "Kiwano (horned melon)"          
 [59] "Kiwifruit"                       "Kumquat"                        
 [61] "Lemon"                           "Lime"                           
 [63] "Loganberry"                      "Longan"                         
 [65] "Loquat"                          "Lychee"                         
 [67] "Mandarine"                       "Mango"                          
 [69] "Mangosteen"                      "Marionberry"                    
 [71] "Melon"                           "Miracle fruit"                  
 [73] "Monstera Delisiousa"             "Mulberry"                       
 [75] "Nance"                           "Nectarine"                      
 [77] "Olive"                           "Orange"                         
 [79] "Papaya"                          "Passionfruit"                   
 [81] "Pea"                             "Peach"                          
 [83] "Pear"                            "Persimmon"                      
 [85] "Pineapple (Ananas)"              "Pineberry"                      
 [87] "Pitaya"                          "Plantain"                       
 [89] "Plum"                            "Plumcot (or Pluot)"             
 [91] "Pomegranate"                     "Pomelo"                         
 [93] "Prune (dried plum)"              "Pumpkin"                        
 [95] "Purple mangosteen"               "Quince"                         
 [97] "Raisin"                          "Rambutan (or Mamin Chino)"      
 [99] "Raspberry"                       "Redcurrant"                     
[101] "Salak"                           "Salal berry"                    
[103] "Salmonberry"                     "Satsuma"                        
[105] "Soursop"                         "Squash"                         
[107] "Star apple"                      "Star fruit"                     
[109] 

In [ ]:
words

[1] "a"           "able"        "about"       "absolute"    "accept"     
  [6] "account"     "achieve"     "across"      "act"         "active"     
 [11] "actual"      "add"         "address"     "admit"       "advertise"  
 [16] "affect"      "afford"      "after"       "afternoon"   "again"      
 [21] "against"     "age"         "agent"       "ago"         "agree"      
 [26] "air"         "all"         "allow"       "almost"      "along"      
 [31] "already"     "alright"     "also"        "although"    "always"     
 [36] "america"     "amount"      "and"         "another"     "answer"     
 [41] "any"         "apart"       "apparent"    "appear"      "apply"      
 [46] "appoint"     "approach"    "appropriate" "area"        "argue"      
 [51] "arm"         "around"      "arrange"     "art"         "as"         
 [56] "ask"         "associate"   "assume"      "at"          "attend"     
 [61] "authority"   "available"   "aware"       "away"        "awful"      
 [66] "baby"        "back"        "bad"         "bag"         "balance"    
 [71] "ball"        "bank"        "bar"         "base"        "basis"      
 [76] "be"          "bear"        "beat"        "beauty"      "because"    
 [81] "become"      "bed"         "before"      "begin"       "behind"     
 [86] "believe"     "benefit"     "best"        "bet"         "between"    
 [91] "big"         "bill"        "birth"       "bit"         "black"      
 [96] "bloke"       "blood"       "blow"        "blue"        "board"      
[101] "boat"        "body"        "book"        "both"        "bother"     
[106] "bottle"      "bottom"      "box"         "boy"         "break"      
[111] "brief"       "brilliant"   "bring"       "britain"     "brother"    
[116] "budget"      "build"       "bus"         "business"    "busy"       
[121] "but"         "buy"         "by"          "cake"        "call"       
[126] "can"         "car"         "card"        "care"        "carry"      
[131] "case"        "cat"         "catch"       "cause"       "cent"       
[136] "centre"      "certain"     "chair"       "chairman"    "chance"     
[141] "change"      "chap"        "character"   "charge"      "cheap"      
[146] "check"       "child"       "choice"      "choose"      "Christ"     
[151] "Christmas"   "church"      "city"        "claim"       "class"      
[156] "clean"       "clear"       "client"      "clock"       "close"      
[161] "closes"      "clothe"      "club"        "coffee"      "cold"       
[166] "colleague"   "collect"     "college"     "colour"      "come"       
[171] "comment"     "commit"      "committee"   "common"      "community"  
[176] "company"     "compare"     "complete"    "compute"     "concern"    
[181] "condition"   "confer"      "consider"    "consult"     "contact"    
[186] "continue"    "contract"    "control"     "converse"    "cook"       
[191] "copy"        "corner"      "correct"     "cost"        "could"      
[196] "council"     "count"       "country"     "county"      "couple"     
[201] "course"      "court"       "cover"       "create"      "cross"      
[206] "cup"         "current"     "cut"         "dad"         "danger"     
[211] "date"        "day"         "dead"        "deal"        "dear"       
[216] "debate"      "decide"      "decision"    "deep"        "definite"   
[221] "degree"      "department"  "depend"      "describe"    "design"     
[226] "detail"      "develop"     "die"         "difference"  "difficult"  
[231] "dinner"      "direct"      "discuss"     "district"    "divide"     
[236] "do"          "doctor"      "document"    "dog"         "door"       
[241] "double"      "doubt"       "down"        "draw"        "dress"      
[246] "drink"       "drive"       "drop"        "dry"         "due"        
[251] "during"      "each"        "early"       "east"        "easy"       
[256] "eat"         "economy"     "educate"     "effect"      "egg"        
[261] "eight"       "either"      "elect"       "e

In [ ]:
length(words)

[1] 980

As you can see above, the word list is 980 words long. Let's use these to do a quick detour on indexing.


In [ ]:
words[1:5]

[1] "a"        "able"     "about"    "absolute" "accept"

In [ ]:
words[1]

[1] "a"

In [ ]:
words[5]

[1] "accept"

The first line above takes the first five elements of the list. R starts at base 1 for indexing. So `words[1]` is "a", and `words[5]` "accept".    

Finally, R also has a dataset called `sentences` that we can make use of.

In [ ]:
length(sentences)

[1] 720

In [ ]:
sentences[1:5]

[1] "The birch canoe slid on the smooth planks." 
[2] "Glue the sheet to the dark blue background."
[3] "It's easy to tell the depth of a well."     
[4] "These days a chicken leg is a rare dish."   
[5] "Rice is often served in round bowls."

# Manipulating strings

We have lots of strings to play with now. You can combine, or “concatenate”, strings very naturally using the "+" sign. Note that in the below we also add in a space between the two strings. You could potentially add anything you want in this particular spot (and can keep going with more strings).

In [ ]:
a_string <- "A first sentence!"
second_string <- "Wow, two sentences."
combined_string <- paste(a_string, second_string, sep = " ")
combined_string

[1] "A first sentence! Wow, two sentences."

We can concatenate *all* the strings using `paste()` with the `collapse` argument.

In [ ]:
paste(month_name, collapse = " then ")

[1] "January then February then March then April then May then June then July then August then September then October then November then December"

Let's work with those months to match abbreviations to month names. There are multiple ways to do this. We could do a `for` loop that iterates through the elements:

In [ ]:
month_explanations <- c()
for(i in 1:12){
    new_string <- paste(month.abb[i], month.name[i], sep = " stands for ")
    month_explanations <- c(month_explanations, new_string)
}
month_explanations

[1] "Jan stands for January"   "Feb stands for February" 
 [3] "Mar stands for March"     "Apr stands for April"    
 [5] "May stands for May"       "Jun stands for June"     
 [7] "Jul stands for July"      "Aug stands for August"   
 [9] "Sep stands for September" "Oct stands for October"  
[11] "Nov stands for November"  "Dec stands for December"

Let's do the same with letter pairs

In [ ]:
letterpairs <- c()
for(i in 1:length(letters_string)){
    letterpair <- paste(letters_string[i], LETTERS_string[i], sep = "")
    letterpairs <- c(letterpairs, letterpair)
}
letterpairs

[1] "aA" "bB" "cC" "dD" "eE" "fF" "gG" "hH" "iI" "jJ" "kK" "lL" "mM" "nN" "oO"
[16] "pP" "qQ" "rR" "sS" "tT" "uU" "vV" "wW" "xX" "yY" "zZ"

You can split up a string into pieces, based on a pattern, with the "split" method.

In [ ]:
str_split(combined_string, "! ")

[[1]]
[1] "A first sentence"    "Wow, two sentences."

## Substrings (Slices)

Substrings are just slices. We are just taking a little slice from a longer string and exporting it. For instance, to get a list of the second through fourth character in each fruit name:

In [ ]:
substringfromfruit <- str_sub(fruit, start = 2, end = 4)
substringfromfruit

[1] "ppl" "pri" "voc" "ana" "ell" "ilb" "lac" "lac" "loo" "lue" "oys" "rea"
[13] "ana" "ant" "her" "her" "hil" "lem" "lou" "oco" "ran" "ucu" "urr" "ams"
[25] "ate" "rag" "uri" "ggp" "lde" "eij" "ig"  "oji" "oos" "rap" "rap" "uav"
[37] "one" "uck" "ack" "amb" "uju" "iwi" "umq" "emo" "ime" "oqu" "ych" "and"
[49] "ang" "ulb" "ect" "ut"  "liv" "ran" "ame" "apa" "ass" "eac" "ear" "ers"
[61] "hys" "ine" "lum" "ome" "ome" "urp" "uin" "ais" "amb" "asp" "edc" "ock"
[73] "ala" "ats" "tar" "tra" "ama" "ang" "gli" "ate"

You can also count backwards from the end of the string in order to extract a slice from that end. To do so, use negative numbers and count backwards. 

In [ ]:
subfromend <- str_sub(fruit, start = -2, end = -1)
print(subfromend)

 [1] "le" "ot" "do" "na" "er" "ry" "ry" "nt" "ge" "ry" "ry" "it" "on" "pe" "ya"
[16] "ry" "er" "ne" "ry" "ut" "ry" "er" "nt" "on" "te" "it" "an" "nt" "ry" "oa"
[31] "ig" "ry" "ry" "pe" "it" "va" "ew" "ry" "it" "ul" "be" "it" "at" "on" "me"
[46] "at" "ee" "ne" "go" "ry" "ne" "ut" "ve" "ge" "lo" "ya" "it" "ch" "ar" "on"
[61] "is" "le" "um" "te" "lo" "en" "ce" "in" "an" "ry" "nt" "on" "ry" "ma" "it"
[76] "ry" "lo" "ne" "it" "on"


You can use slicing to extract data from strings:

In [ ]:
some_dates <- c("1999/01/01","1998/12/15","2001/09/03")
years <- str_sub(some_dates, start = 1, end = 4)
years

[1] "1999" "1998" "2001"

In [ ]:
months <- str_sub(some_dates, start = 6, end = 7)
print(months)

[1] "01" "12" "09"


Getting a copy of a string with specific positions replaced is also a matter of slicing:

In [ ]:
apple <- "apple"
zebra <- "--!ZEBRA!--"
zebraapple <- paste(str_sub(apple, 1, 2), zebra, str_sub(apple, start = 3), sep = "")
zebraapple

[1] "ap--!ZEBRA!--ple"

## Capitalization

Strings have a simple casefolding method that can be applied:

In [ ]:
tolower(combined_string)

[1] "a first sentence! wow, two sentences."

In [ ]:
toupper(combined_string)

[1] "A FIRST SENTENCE! WOW, TWO SENTENCES."

## White Space

R has several methods to trim excess white space *off the ends* of strings:

In [ ]:
lotsofspace = '   Why   so much  space?   '
trimws(lotsofspace, which = "left") 

[1] "Why   so much  space?   "

In [ ]:
trimws(lotsofspace, which = "right")

[1] "   Why   so much  space?"

In [ ]:
trimws(lotsofspace, which = "both")

[1] "Why   so much  space?"

## Matching substrings

If we're looking for specific substrings, there are string methods to do that.

In [ ]:
str_locate("strawberry", "berry")

start end
[1,] 6     10

That returns a matrix with the position of the beginning and the end of the match. If there is no match, it returns an empty matrix.

In [ ]:
str_locate("apple", "berry")

start end
[1,] NA    NA

If there are multiple matches, `str_locate()` returns the position of the first match.

In [ ]:
str_locate("berryberryboberrybananafanafoferrymemymomerry berry", "berry")

start end
[1,] 1     5

You can pull all of the matches with `str_locate_all()`

In [ ]:
str_locate_all("berryberryboberrybananafanafoferrymemymomerry berry", "berry")

[[1]]
     start end
[1,]     1   5
[2,]     6  10
[3,]    13  17
[4,]    47  51

We can print all the matches from a vector of strings using `str_detect()` and relying on indexing `[]`

In [ ]:
fruit[str_detect(fruit, "berry")]

[1] "bilberry"    "blackberry"  "blueberry"   "boysenberry" "cloudberry" 
 [6] "cranberry"   "elderberry"  "goji berry"  "gooseberry"  "huckleberry"
[11] "mulberry"    "raspberry"   "salal berry" "strawberry"

We can get a copy of the string with the substring replaced with something else:

In [ ]:
str_replace("strawberry","berry", "fish")

[1] "strawfish"

In [ ]:
fishfruit <- str_replace(fruit, "berry","fish")
fishfruit

[1] "apple"             "apricot"           "avocado"          
 [4] "banana"            "bell pepper"       "bilfish"          
 [7] "blackfish"         "blackcurrant"      "blood orange"     
[10] "bluefish"          "boysenfish"        "breadfruit"       
[13] "canary melon"      "cantaloupe"        "cherimoya"        
[16] "cherry"            "chili pepper"      "clementine"       
[19] "cloudfish"         "coconut"           "cranfish"         
[22] "cucumber"          "currant"           "damson"           
[25] "date"              "dragonfruit"       "durian"           
[28] "eggplant"          "elderfish"         "feijoa"           
[31] "fig"               "goji fish"         "goosefish"        
[34] "grape"             "grapefruit"        "guava"            
[37] "honeydew"          "hucklefish"        "jackfruit"        
[40] "jambul"            "jujube"            "kiwi fruit"       
[43] "kumquat"           "lemon"             "lime"             
[46] "loquat"            "lychee"            "mandarine"        
[49] "mango"             "mulfish"           "nectarine"        
[52] "nut"               "olive"             "orange"           
[55] "pamelo"            "papaya"            "passionfruit"     
[58] "peach"             "pear"              "persimmon"        
[61] "physalis"          "pineapple"         "plum"             
[64] "pomegranate"       "pomelo"            "purple mangosteen"
[67] "quince"            "raisin"            "rambutan"         
[70] "raspfish"          "redcurrant"        "rock melon"       
[73] "salal fish"        "satsuma"           "star fruit"       
[76] "strawfish"         "tamarillo"         "tangerine"        
[79] "ugli fruit"        "watermelon"

# Regular Expressions

Now we can try to generalize beyond the simple word case to broader patterns. Regular expressions allow us to capture more complex phenomena, which are often what we need when we are doing work with text-as-data. 

To illustrate, we'll go back to the `sentences` data from earlier.  Let's look for the substring "dog" in it.

In [ ]:
str_subset(sentences, "dog")

[1] "Cats and dogs each hate the other."        
[2] "The child almost hurt the small dog."      
[3] "The kitten chased the dog down the street."
[4] "Tend the sheep while the dog wanders."     
[5] "The sheep were led home by a dog."         
[6] "Five years he lived with a shaggy dog."

That's just a search for a substring. Let's start to expand from there and use regex to look for more complex patterns than just substrings.

#### Square brackets for “or” (disjunction) of characters.

Match “any one of” the characters in the square brackets.

In [ ]:
# note that the spaces are important within the quotes below. what happens if you remove them? Why?
str_subset(sentences, ' [hbd]og ')

[1] "The hog crawled under the high fence."     
[2] "The kitten chased the dog down the street."
[3] "Tend the sheep while the dog wanders."

#### Square brackets with ^ for negation.

Match “anything but one of” the characters in the square brackets.

In [ ]:
# look for anything but hbd
str_subset(sentences, ' [^hbd]og ')


[1] "Split the log with a quick, sharp blow."
[2] "Watch the log float in the wide river."

#### Square brackets for “or” over a range of characters

 Or we can use `-` to look across a range of characters

In [ ]:
# look for anything between b and l
str_subset(sentences, ' [b-l]ot ')


#### Pipe operator for "or" over multi-character patterns

When we need an “or” over multi-character patterns, we can use the “pipe” operator, using parentheses as necessary to identify what’s with what. Note that we *should not* use this with the square brackets now. Instead, we use the square brackets to put an option for upper or lower case s at the beginning of the regex match.

In [ ]:
str_subset(sentences, '([Ss]pl|f)it ')

[1] "Split the log with a quick, sharp blow."
[2] "A ripe plum is fit for a king's palate."
[3] "The key you designed will fit the lock."
[4] "Ripe pears are fit for a queen's table."
[5] "The gloss on top made it unfit to read."

#### Special characters and the backslash

In addition to the backslash itself, there are several characters that have special meaning in R regexes, and (may) have to be escaped in order to match the literal character. Here are the big ones: ^ $ . * + | ! ? ( ) [ ] { } < >.

For example, the period – “.” – means “any character but a newline.” It’s a wildcard. We get different results when we escape or don’t escape it.

In [ ]:
allchars <- str_extract_all(combined_string, ".")
allchars

[[1]]
 [1] "A" " " "f" "i" "r" "s" "t" " " "s" "e" "n" "t" "e" "n" "c" "e" "!" " " "W"
[20] "o" "w" "," " " "t" "w" "o" " " "s" "e" "n" "t" "e" "n" "c" "e" "s" "."

In [ ]:
allchars <- str_extract_all(combined_string, "\\.")
allchars

[[1]]
[1] "."

In [ ]:
allchars <- str_extract_all(combined_string, "s.")
allchars

[[1]]
[1] "st" "se" "se" "s."

In [ ]:
allchars <- str_extract_all(combined_string, "s\\.")
allchars

[[1]]
[1] "s."

Some of these are only special characters in certain contexts and don’t have to be escaped to be recognized when not in those contexts. But they can be escaped in all circumstances and I recommend that rather than trying to figure out the exact rules.

The exclamation point is such a character.

In [ ]:
allchars <- str_extract_all(combined_string, "\\!")
allchars

[[1]]
[1] "!"

#### Class shorthands

Conversely, there are a number of characters that have special meaning only when escaped. The main ones for now are “\\w” (any alphanumeric character), “\\s” (any space character), and “\\d” (any numeric digit). The capitalized versions of these are used to mean “anything but” that class.

In [ ]:
allchars <- str_extract_all(combined_string, "\\w")
allchars

[[1]]
 [1] "A" "f" "i" "r" "s" "t" "s" "e" "n" "t" "e" "n" "c" "e" "W" "o" "w" "t" "w"
[20] "o" "s" "e" "n" "t" "e" "n" "c" "e" "s"

In [ ]:
allchars <- str_extract_all(combined_string, "\\W")
allchars

[[1]]
[1] " " " " "!" " " "," " " " " "."

In [ ]:
allchars <- str_extract_all(combined_string, "\\s")
allchars

[[1]]
[1] " " " " " " " " " "

In [ ]:
allchars <- str_extract_all(combined_string, "\\S")
allchars

[[1]]
 [1] "A" "f" "i" "r" "s" "t" "s" "e" "n" "t" "e" "n" "c" "e" "!" "W" "o" "w" ","
[20] "t" "w" "o" "s" "e" "n" "t" "e" "n" "c" "e" "s" "."

In [ ]:
allchars <- str_extract_all(combined_string, "\\d")
allchars

[[1]]
character(0)

In [ ]:
allchars <- str_extract_all(combined_string, "\\D")
allchars

[[1]]
 [1] "A" " " "f" "i" "r" "s" "t" " " "s" "e" "n" "t" "e" "n" "c" "e" "!" " " "W"
[20] "o" "w" "," " " "t" "w" "o" " " "s" "e" "n" "t" "e" "n" "c" "e" "s" "."

#### Quantifiers: * (zero or more of the previous)

This is also known as the “Kleene star” (pronounced clean-ee), after its original user (Kleene) who introduced the notation in formal logic.

In [ ]:
allchars <- str_extract_all(combined_string, "\\d*") # any string of 0 or more digits
allchars

[[1]]
 [1] "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" ""
[26] "" "" "" "" "" "" "" "" "" "" "" "" ""

Note the "zero" or more led it to identify every position of the string as a match, many of them empty (containing no characters).

#### Quantifiers: + (one or more of the previous)

This is also known as the “Kleene plus.”

In [ ]:
allchars <- str_extract_all(combined_string, "\\d+")
allchars

[[1]]
character(0)

#### Quantifiers {n} {n,m} and {n,}

{n} = “exactly n” of the previous
{n,m} = “between n and m” of the previous
{n,} = “n or more” of the previous

In [ ]:
xstring <- c("x xx xxx xxxx xxxxx")

str_extract_all(xstring, "x{3}") # 3 x's


[[1]]
[1] "xxx" "xxx" "xxx"

In [ ]:
str_extract_all(xstring, "x{3,4}") # 3 or 4 x's

[[1]]
[1] "xxx"  "xxxx" "xxxx"

In [ ]:
str_extract_all(xstring, "x{3,}") # 3 or more x's

[[1]]
[1] "xxx"   "xxxx"  "xxxxx"

#### Quantifier ? (zero or one of the previous)

In [ ]:
str_extract_all(combined_string, "\\d?")

[[1]]
 [1] "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" ""
[26] "" "" "" "" "" "" "" "" "" "" "" "" ""

In [ ]:
str_subset(sentences, " [bp]?eat ")

[1] "The heart beat strongly and with firm strokes."
[2] "Burn peat after the logs give out."            
[3] "A speedy man can beat this track mark."        
[4] "Even the worst will beat his low score."       
[5] "Quench your thirst, then eat the crackers."

#### Anchors at beginning and end of string

In [ ]:
# beginning of string
str_extract_all(combined_string, "^.")

[[1]]
[1] "A"

In [ ]:
# end of string
str_extract_all(combined_string, ".$")

[[1]]
[1] "."

#### Anchors at word boundaries

Similarly, we can identify "word boundaries" with \b. 

In [ ]:
# word boundary
str_extract_all(xstring, "\\b\\w+?\\b")

[[1]]
[1] "x"     "xx"    "xxx"   "xxxx"  "xxxxx"

In [ ]:
# word boundary
str_extract_all(combined_string, "\\b\\w+?\\b")

[[1]]
[1] "A"         "first"     "sentence"  "Wow"       "two"       "sentences"

## An example

Consider the following text from a Congressional Bill. We're going to try to use regular expressions to make data out of the appropriations dollars and purposes in bullets 1-9.

In [ ]:
text = "SEC. 101. FISCAL YEAR 2017.
(a) In General.--There are authorized to be appropriated to NASA
for fiscal year 2017 $19,508,000,000, as follows:
(1) For Exploration, $4,330,000,000.
(2) For Space Operations, $5,023,000,000.
(3) For Science, $5,500,000,000.
(4) For Aeronautics, $640,000,000.
(5) For Space Technology, $686,000,000.
(6) For Education, $115,000,000.
(7) For Safety, Security, and Mission Services,
$2,788,600,000.
(8) For Construction and Environmental Compliance and
Restoration, $388,000,000.
(9) For Inspector General, $37,400,000.
(b) Exception.--In addition to the amounts authorized to be
appropriated for each account under subsection (a), there are
authorized to be appropriated additional funds for each such account,
but only if the authorized amounts for all such accounts are fully
provided for in annual appropriation Acts, consistent with the
discretionary spending limits in section 251(c) of the Balanced Budget
and Emergency Deficit Control Act of 1985."

Note that's *one* string with a bunch of newline characters.

In [ ]:
text

[1] "SEC. 101. FISCAL YEAR 2017.\n(a) In General.--There are authorized to be appropriated to NASA\nfor fiscal year 2017 $19,508,000,000, as follows:\n(1) For Exploration, $4,330,000,000.\n(2) For Space Operations, $5,023,000,000.\n(3) For Science, $5,500,000,000.\n(4) For Aeronautics, $640,000,000.\n(5) For Space Technology, $686,000,000.\n(6) For Education, $115,000,000.\n(7) For Safety, Security, and Mission Services,\n$2,788,600,000.\n(8) For Construction and Environmental Compliance and\nRestoration, $388,000,000.\n(9) For Inspector General, $37,400,000.\n(b) Exception.--In addition to the amounts authorized to be\nappropriated for each account under subsection (a), there are\nauthorized to be appropriated additional funds for each such account,\nbut only if the authorized amounts for all such accounts are fully\nprovided for in annual appropriation Acts, consistent with the\ndiscretionary spending limits in section 251(c) of the Balanced Budget\nand Emergency Deficit Control Act of 1985."

Lets play around with a few things. Extract all contiguous sequences of one or more numbers.

In [ ]:
digitmatches <- str_extract_all(text, '[0-9]+') # one or more consecutive digits
digitmatches

[[1]]
 [1] "101"  "2017" "2017" "19"   "508"  "000"  "000"  "1"    "4"    "330" 
[11] "000"  "000"  "2"    "5"    "023"  "000"  "000"  "3"    "5"    "500" 
[21] "000"  "000"  "4"    "640"  "000"  "000"  "5"    "686"  "000"  "000" 
[31] "6"    "115"  "000"  "000"  "7"    "2"    "788"  "600"  "000"  "8"   
[41] "388"  "000"  "000"  "9"    "37"   "400"  "000"  "251"  "1985"

That does two things we don't like ... separates numbers at the 1000s separating comma and gets numbers ("101", "2017", etc.) that aren't dollar amounts. So, let's try getting everything that:
* Starts with a "$" (which needs to be escaped)
* Followed by one or more strings of commas or digits.

In [ ]:
dollarmatches <- str_extract_all(text, '\\$[,0-9]+') # $ followed by one or more digits or commas
dollarmatches

[[1]]
 [1] "$19,508,000,000," "$4,330,000,000"   "$5,023,000,000"   "$5,500,000,000"  
 [5] "$640,000,000"     "$686,000,000"     "$115,000,000"     "$2,788,600,000"  
 [9] "$388,000,000"     "$37,400,000"

Almost ... don't like that extra comma on the first number. Let's require it to end with a number.

In [ ]:
dollarmatches <- str_extract_all(text, '\\$[,0-9]+[0-9]') # $ followed by one or more digits or commas and ends in a number
dollarmatches

[[1]]
 [1] "$19,508,000,000" "$4,330,000,000"  "$5,023,000,000"  "$5,500,000,000" 
 [5] "$640,000,000"    "$686,000,000"    "$115,000,000"    "$2,788,600,000" 
 [9] "$388,000,000"    "$37,400,000"

The things we want are demarcated by numbered items in parentheses. Let's see if we can extract those:

In [ ]:
bulletmatches <- str_extract_all(text, '\\([0-9]+\\)') # ( followed by a digit followed by )
bulletmatches



[[1]]
[1] "(1)" "(2)" "(3)" "(4)" "(5)" "(6)" "(7)" "(8)" "(9)"

Let's go back to the original and get rid of the newlines. Note that the string.replace() method doesn't accept regular expressions and you need to use re.sub().

In [ ]:
one_line <- str_replace(text, "\n", " ")
one_line

[1] "SEC. 101. FISCAL YEAR 2017. (a) In General.--There are authorized to be appropriated to NASA\nfor fiscal year 2017 $19,508,000,000, as follows:\n(1) For Exploration, $4,330,000,000.\n(2) For Space Operations, $5,023,000,000.\n(3) For Science, $5,500,000,000.\n(4) For Aeronautics, $640,000,000.\n(5) For Space Technology, $686,000,000.\n(6) For Education, $115,000,000.\n(7) For Safety, Security, and Mission Services,\n$2,788,600,000.\n(8) For Construction and Environmental Compliance and\nRestoration, $388,000,000.\n(9) For Inspector General, $37,400,000.\n(b) Exception.--In addition to the amounts authorized to be\nappropriated for each account under subsection (a), there are\nauthorized to be appropriated additional funds for each such account,\nbut only if the authorized amounts for all such accounts are fully\nprovided for in annual appropriation Acts, consistent with the\ndiscretionary spending limits in section 251(c) of the Balanced Budget\nand Emergency Deficit Control Act of 1985."

and find all the matches from "(number)" to a period, lazily rather than greedily:

In [ ]:
item_strings = str_extract_all(one_line, "\\(\\d\\).+?\\.")
item_strings

[[1]]
[1] "(1) For Exploration, $4,330,000,000."     
[2] "(2) For Space Operations, $5,023,000,000."
[3] "(3) For Science, $5,500,000,000."         
[4] "(4) For Aeronautics, $640,000,000."       
[5] "(5) For Space Technology, $686,000,000."  
[6] "(6) For Education, $115,000,000."         
[7] "(9) For Inspector General, $37,400,000."  



We can use a capture group to gather just the "for what" data ...

In [ ]:
for_matches <- str_match(item_strings[[1]], '(For )(.+)(, \\$)')
for_matches

[,1]                     [,2] [,3]              [,4]
[1,] For Exploration, $       For  Exploration       , $ 
[2,] For Space Operations, $  For  Space Operations  , $ 
[3,] For Science, $           For  Science           , $ 
[4,] For Aeronautics, $       For  Aeronautics       , $ 
[5,] For Space Technology, $  For  Space Technology  , $ 
[6,] For Education, $         For  Education         , $ 
[7,] For Inspector General, $ For  Inspector General , $

We can pull the type then from the third column. 

In [ ]:
for_matches[,3]

[1] "Exploration"       "Space Operations"  "Science"          
[4] "Aeronautics"       "Space Technology"  "Education"        
[7] "Inspector General"

We can also use a capture group just for the money data

In [ ]:
money_matches <- str_match(item_strings[[1]], '\\$([, \\d]+)')
money_matches[,2]

[1] "4,330,000,000" "5,023,000,000" "5,500,000,000" "640,000,000"  
[5] "686,000,000"   "115,000,000"   "37,400,000"

We'll probably want those just to be numbers, so we'll strip the commas:

In [ ]:
money_strings_clean <- str_replace_all(money_matches[,2], ",", "")
money_strings_clean

[1] "4330000000" "5023000000" "5500000000" "640000000"  "686000000" 
[6] "115000000"  "37400000"

We'll turn that into numbers and reformat them as millions.

In [ ]:
millions <- as.numeric(money_strings_clean)/10e6
millions

[1] 433.00 502.30 550.00  64.00  68.60  11.50   3.74

Finally, we can format the data in a dataframe for later processing.

In [ ]:
myDf <- as_tibble(as.matrix(cbind(for_matches[,3],millions)))

In [ ]:
colnames(myDf) <- c("For", "Millions")
myDf

For               Millions
1 Exploration       433     
2 Space Operations  502.3   
3 Science           550     
4 Aeronautics       64      
5 Space Technology  68.6    
6 Education         11.5    
7 Inspector General 3.74